### DISCUSSION TITLE
MSW handlers and JSDOM can be very slow

### DISCUSSION THREAD

**Discussion by tkrotoff** [mswjs/msw#2141] [Show and tell]
JSDOM/whatwg-url is slow to parse URLs with long query string.

https://github.com/jsdom/jsdom/issues/3712

This repo is an example with 500 MSW handlers: https://github.com/tkrotoff/jsdom-slow-URL-parsing

Results (MacBook Pro 16" M1, macOS 14.4.1, Node.js v20.12.1):

- Vitest `environment: 'node'`
  - JSDOM + MSW short URL: 23.951ms
  - JSDOM + MSW long URL: 27.046ms

- Vitest `environment: 'happy-dom'`
  - JSDOM + MSW short URL: 20.962ms
  - JSDOM + MSW long URL: 29.937ms

- Vitest `environment: 'jsdom'`
  - JSDOM + MSW short URL: 66.226ms
  - JSDOM + MSW long URL: **3.740s**

- Vitest `environment: 'jsdom'` + whatwg-url patched
  - JSDOM + MSW short URL: 34.554ms
  - JSDOM + MSW long URL: **126.326ms**
  
Here is the dirty patch:

```DIFF
diff --git a/node_modules/whatwg-url/lib/URL-impl.js b/node_modules/whatwg-url/lib/URL-impl.js
index eb66cae..1042729 100644
--- a/node_modules/whatwg-url/lib/URL-impl.js
+++ b/node_modules/whatwg-url/lib/URL-impl.js
@@ -244,3 +244,13 @@ exports.implementation = class URLImpl {
     this._url.path = this._url.path.replace(/\u0020+$/u, "");
   }
 };
+
+class URLImpl extends URL {
+  constructor(globalObject, constructorArgs) {
+    const url = constructorArgs[0];
+    const base = constructorArgs[1];
+    super(url, base);
+  }
+}
+
+exports.implementation = URLImpl;
diff --git a/node_modules/whatwg-url/lib/URLSearchParams-impl.js b/node_modules/whatwg-url/lib/URLSearchParams-impl.js
index 934c4ff..f4c3305 100644
--- a/node_modules/whatwg-url/lib/URLSearchParams-impl.js
+++ b/node_modules/whatwg-url/lib/URLSearchParams-impl.js
@@ -137,3 +137,12 @@ exports.implementation = class URLSearchParamsImpl {
     return urlencoded.serializeUrlencoded(this._list);
   }
 };
+
+class URLSearchParamsImpl extends URLSearchParams {
+  constructor(globalObject, constructorArgs) {
+    const init = constructorArgs[0];
+    super(init);
+  }
+}
+
+exports.implementation = URLSearchParamsImpl;
```

<br><br>

On our internal application (react-testing-library, jsdom, Vitest, MSW), we have seen tremendous performance increase on our CI thanks to this dirty patch.

**Comment by kettanaito** [comment]
Hi, @tkrotoff. There's nothing MSW can or should do about this. We are not doing URL parsing, it's something JSDOM (`whatwg-url` transitively) does for you. Please raise this issue in the respective repositories. Thanks!

**Reply by tkrotoff** [reply]
Of course. That's why this is a "discussion" in category "Show and tell"

I should have been more precise: it is just for information if someone encounters the same issue.
To avoid spending hours of debugging, understanding why there is a performance issue.

This issue has already been reported to JSDOM/whatwg-url. The author doesn't want to fix it because _jsdom is a project meant to implement specifications from scratch for fun. Performance is a secondary objective for us._
So I advice anyone to use Happy DOM instead: it has more potential IMHO.
